man/man3/bash-plop.3
.TH	BASH-PLOP 3 2021-08-07
.SH NAME
bash-plop
-
A bash testing framework suitable for 42 Projects.
.SH	SYNOPSIS
.nf
.B	plop_init
.fi
.nf
.B	plop_end
.fi
.nf
.BI	"plop_commands_exist " commands
.fi
.nf
.BI	"plop_test_command " command
.fi
.nf
.B	plop_test_summary
.fi
.nf
.BI	"plop_pipe_cmd " command1 " " command2
.fi
.SH	DESCRIPTION
.TP
.B	plop_init
This function should be called first. It initializes all needed variables.
.TP
.B	plop_end
This function should be called at the end of all tests.
It prints a summary of all tests.
.TP
.B	plop_commands_exist
This function check if the commands given in parameters as an array exist.
If not, the test exits immediately.
.TP
.B	plop_test_command
This function prints a \fIdescription\fR
and executes the \fIcommand\fR passed in argument.
The \fIdescription\fR can be passed using the variable \fBPLOP_DESCRIPTION\fR.
If the user defined a \fIplop_setup\fR function, it will be called first.
The function returns the exit status of the executed command,
and also stores it in the variable \fBPLOP_EXIT_STATUS\fR.
If the test times out, the variable \fBPLOP_TIMED_OUT\fR will be set to 1.
You can set a prefix command in the variable \fBPLOP_LEAK_CMD\fR:
it is called like that because it can mainly used to check memory leaks.
You can also decide to skip the test
by setting the variable \fBPLOP_SKIP\fR to 1:
the main purpose is to only print the test description
without actually executing the test suite.
.TP
.B	plop_test_summary
This function prints a summary of the previously executed command
(if it failed or not).
If the user defined a \fIplop_teardown\fR function,
it will be called at the end.
If the variable \fBPLOP_SKIP\fR is set to 1, you should still call this function
to reset the test and print a newline.
.TP
.B	plop_pipe_cmd
This function allows you to pipe \fIcommand1\fR into \fIcommand2\fR.
It is useful when you want to use bash pipes in the test,
because it is not possible to store that in a string.
.SH	ENVIRONMENT
.TP
.B	PLOP_DESCRIPTION
The description displayed for the test.
.TP
.B	PLOP_EXIT_STATUS
The exit status of the tested command.
.TP
.B	PLOP_LEAK_CMD
The prefix command in front of the the tested one.
The main purpose is to use a program like \fIvalgrind\fR to check leaks.
.TP
.B	PLOP_LINE_LENGTH
The length of the description line. By default, it is set to 80.
.TP
.B	PLOP_MIN_NUM_LENGHT
The minimum length of the displayed test number on the description line.
By default, it is set to 2.
.TP
.B	PLOP_RESULT_COLOR
The color (in ANSI escape codes) in which the test result will be outputed.
By default, if the variable \fBPLOP_TEST_RESULT\fR is set to "OK",
the color is set to green, otherwise it is set to red.
.TP
.B	PLOP_SKIP
If set to 1, the test will be skipped. Only the description will be displayed.
.TP
.B	PLOP_SKIP_LEAKS
If set to 1, the leak test will be skipped.
.TP
.B	PLOP_TEST_NUM
The number of the test (automatically incremented in \fIplop_test_summary\fR).
.TP
.B	PLOP_TEST_OUTPUT
The file in which the test output will be redirected (/dev/null by default).
.TP
.B	PLOP_TEST_RESULT
The test result. It should be one of these "OK", "KO",
"LK" (leak) or "TO" (timeout).
.TP
.B	PLOP_TIMED_OUT
1 if the test has timed out.
.TP
.B	PLOP_TIMEOUT_SECONDS
The time in seconds after which the test is killed.
If set to 0, the test will never be killed.
By default, it is set to 5.


bash-plop.sh
#!/bin/bash

RESET_COLOR="\033[0m"
BOLD="\033[1m"
UNDERLINE="\033[4m"
RED="\033[31m"
GREEN="\033[32m"
YELLOW="\033[33m"
BLUE="\033[34m"
MAGENTA="\033[35m"

plop_fatal_error()
{
	if [ -z "$1" ]
	then
		message="fatal error"
	else
		message="$1"
	fi
	if [ -z "$2" ]
	then
		exit_status=1
	else
		exit_status=$2
	fi
	printf "${RED}$message${RESET_COLOR}\n"
	exit $exit_status
}

plop_init()
{
	PLOP_INIT=1
	TESTS_OK=0
	TESTS_KO=0
	TESTS_LK=0
	TESTS_TO=0
	PLOP_TEST_NUM=1
	PLOP_TEST_RESULT="OK"
	PLOP_TEST_INPUT="/dev/null"
	PLOP_TEST_OUTPUT="/dev/null"
	PLOP_LINE_LENGTH=80
	PLOP_MIN_NUM_LENGTH=2
	PLOP_LEAK_CMD=""
	PLOP_SKIP=0
	PLOP_SKIP_LEAKS=0
	PLOP_TIMEOUT_SECONDS=5
	PLOP_TIMED_OUT=0
	exec 3>&1
}

plop_end()
{
	if [ -z "$PLOP_INIT" ] || [ $PLOP_INIT -eq 0 ]
	then
		plop_fatal_error "bash-plop not initialized..."
	fi
	exec 1>&3
	printf "\n"
	[ $TESTS_OK -gt 0 ] && printf "${GREEN}$TESTS_OK OK${RESET_COLOR}"
	[ $TESTS_OK -gt 0 ] && [ $TESTS_KO -gt 0 ] && printf " - "
	[ $TESTS_KO -gt 0 ] && printf "${RED}$TESTS_KO KO${RESET_COLOR}"
	([ $TESTS_OK -gt 0 ] || [ $TESTS_KO -gt 0 ]) && [ $TESTS_LK -gt 0 ] && printf " - "
	[ $TESTS_LK -gt 0 ] && printf "${RED}$TESTS_LK LK${RESET_COLOR}"
	([ $TESTS_OK -gt 0 ] || [ $TESTS_KO -gt 0 ] || [ $TESTS_LK -gt 0 ]) && [ $TESTS_TO -gt 0 ] && printf " - "
	[ $TESTS_TO -gt 0 ] && printf "${RED}$TESTS_TO TO${RESET_COLOR}"
	printf "\n\n"
	if [ $TESTS_KO -eq 0 ] && [ $TESTS_LK -eq 0 ] && [ $TESTS_TO -eq 0 ]
	then
		exit 0
	else
		exit 1
	fi
}

plop_commands_exist()
{
	for cmd in "$@"
	do
		if ! command -v $cmd > /dev/null 2>&1
		then
			plop_fatal_error "command '$cmd' is not installed..."
		fi
	done
	return 0
}

plop_wait_for_timeout()
{
	sleep $PLOP_TIMEOUT_SECONDS
	if kill -0 $1 > /dev/null 2>&1
	then
		touch timed.out > /dev/null 2>&1
		kill $1
	fi
}

plop_timeout_test()
{
	if [ $PLOP_TIMEOUT_SECONDS -gt 0 ]
	then
		"$@" < $PLOP_TEST_INPUT &
		bg_process=$!
		plop_wait_for_timeout $bg_process &
		wait $bg_process
	else
		"$@" < $PLOP_TEST_INPUT
	fi
	PLOP_EXIT_STATUS=$?
	if [ -f timed.out ]
	then
		PLOP_TIMED_OUT=1
		rm -f timed.out > /dev/null 2>&1
	fi
	if ! [ -z "$PLOP_TIMED_OUT" ] && [ $PLOP_TIMED_OUT -gt 0 ]
	then
		return 1
	else
		return 0
	fi
}

plop_test_command()
{
	if [ -z "$PLOP_INIT" ] || [ $PLOP_INIT -eq 0 ]
	then
		plop_fatal_error "bash-plop not initialized..."
	fi
	printf "${BLUE}# %0*d: %-*s  []${RESET_COLOR}" $PLOP_MIN_NUM_LENGTH $PLOP_TEST_NUM $(($PLOP_LINE_LENGTH - 9 - $PLOP_MIN_NUM_LENGTH)) "$PLOP_DESCRIPTION"
	command -v plop_setup > /dev/null 2>&1 && plop_setup
	PLOP_EXIT_STATUS=0
	if ([ -z "$PLOP_SKIP" ] || [ $PLOP_SKIP -eq 0 ]) && [ $# -gt 0 ]
	then
		if [ -z "$PLOP_SKIP_LEAKS" ] || [ "$PLOP_SKIP_LEAKS" -eq 0 ]
		then
			plop_timeout_test $PLOP_LEAK_CMD "$@" > $PLOP_TEST_OUTPUT 2>&1
		else
			plop_timeout_test "$@" > $PLOP_TEST_OUTPUT 2>&1
		fi
	fi
	return $PLOP_EXIT_STATUS
}

plop_test_summary()
{
	if [ -z "$PLOP_SKIP" ] || [ $PLOP_SKIP -eq 0 ]
	then
		if [ -z "$PLOP_TIMED_OUT" ] || [ $PLOP_TIMED_OUT -eq 0 ]
		then
			if [ -z "$PLOP_RESULT_COLOR" ]
			then
				if [ -z "$PLOP_TEST_RESULT" ] || [ "$PLOP_TEST_RESULT" = "OK" ]
				then
					PLOP_RESULT_COLOR=$GREEN
				else
					PLOP_RESULT_COLOR=$RED
				fi
			fi
			case $PLOP_TEST_RESULT in
				"OK")
					TESTS_OK=$(($TESTS_OK + 1));;
				"KO")
					TESTS_KO=$(($TESTS_KO + 1));;
				"LK")
					TESTS_LK=$(($TESTS_LK + 1));;
			esac
		else
			TESTS_TO=$(($TESTS_TO + 1))
			PLOP_TEST_RESULT="TO"
			PLOP_RESULT_COLOR=$RED
		fi
		printf "\r${PLOP_RESULT_COLOR}# %0*d: %-*s [%s]\n${RESET_COLOR}" $PLOP_MIN_NUM_LENGTH $PLOP_TEST_NUM $(($PLOP_LINE_LENGTH - 9 - $PLOP_MIN_NUM_LENGTH)) "$PLOP_DESCRIPTION" "$PLOP_TEST_RESULT"
	else
		printf "\n"
	fi
	command -v plop_teardown > /dev/null 2>&1 && plop_teardown
	PLOP_TEST_NUM=$(($PLOP_TEST_NUM + 1))
	PLOP_TEST_RESULT="OK"
	PLOP_TEST_INPUT="/dev/null"
	PLOP_TEST_OUTPUT="/dev/null"
	PLOP_DESCRIPTION=""
	PLOP_SKIP=0
	PLOP_SKIP_LEAKS=0
	PLOP_TIMED_OUT=0
	PLOP_TIMEOUT_SECONDS=5
	PLOP_RESULT_COLOR=""
}

plop_pipe_cmd()
{
	cmd=$1
	shift 1
	echo $@ | $cmd
}

plop_commands_exist 'touch' 'rm' 'sleep'
